url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/drafts/Draft_76.java

```diff
@@ -0,0 +1,242 @@
+import minewebsocket.org.java_websocket.handshake.HandshakeBuilder;
+import minewebsocket.org.java_websocket.handshake.Handshakedata;
+import minewebsocket.org.java_websocket.handshake.ServerHandshake;
+import minewebsocket.org.java_websocket.handshake.ServerHandshakeBuilder;
+
+public class Draft_76 extends Draft_75 {
+	private boolean failed = false;
+	private static final byte[] closehandshake = { -1, 0 };
+	
+	private final Random reuseableRandom = new Random();
+	
+
+	public static byte[] createChallenge( String key1, String key2, byte[] key3 ) throws InvalidHandshakeException {
+		byte[] part1 = getPart( key1 );
+		byte[] part2 = getPart( key2 );
+		byte[] challenge = new byte[ 16 ];
+		challenge[ 0 ] = part1[ 0 ];
+		challenge[ 1 ] = part1[ 1 ];
+		challenge[ 2 ] = part1[ 2 ];
+		challenge[ 3 ] = part1[ 3 ];
+		challenge[ 4 ] = part2[ 0 ];
+		challenge[ 5 ] = part2[ 1 ];
+		challenge[ 6 ] = part2[ 2 ];
+		challenge[ 7 ] = part2[ 3 ];
+		challenge[ 8 ] = key3[ 0 ];
+		challenge[ 9 ] = key3[ 1 ];
+		challenge[ 10 ] = key3[ 2 ];
+		challenge[ 11 ] = key3[ 3 ];
+		challenge[ 12 ] = key3[ 4 ];
+		challenge[ 13 ] = key3[ 5 ];
+		challenge[ 14 ] = key3[ 6 ];
+		challenge[ 15 ] = key3[ 7 ];
+		MessageDigest md5;
+		try {
+			md5 = MessageDigest.getInstance( "MD5" );
+		} catch ( NoSuchAlgorithmException e ) {
+			throw new RuntimeException( e );
+		}
+		return md5.digest( challenge );
+	}
+
+	private static String generateKey() {
+		Random r = new Random();
+		long maxNumber = 4294967295L;
+		long spaces = r.nextInt( 12 ) + 1;
+		int max = new Long( maxNumber / spaces ).intValue();
+		max = Math.abs( max );
+		int number = r.nextInt( max ) + 1;
+		long product = number * spaces;
+		String key = Long.toString( product );
+		// always insert atleast one random character
+		int numChars = r.nextInt( 12 ) + 1;
+		for( int i = 0 ; i < numChars ; i++ ) {
+			int position = r.nextInt( key.length() );
+			position = Math.abs( position );
+			char randChar = (char) ( r.nextInt( 95 ) + 33 );
+			// exclude numbers here
+			if( randChar >= 48 && randChar <= 57 ) {
+				randChar -= 15;
+			}
+			key = new StringBuilder( key ).insert( position, randChar ).toString();
+		}
+		for( int i = 0 ; i < spaces ; i++ ) {
+			int position = r.nextInt( key.length() - 1 ) + 1;
+			position = Math.abs( position );
+			key = new StringBuilder( key ).insert( position, "\u0020" ).toString();
+		}
+		return key;
+	}
+
+	private static byte[] getPart( String key ) throws InvalidHandshakeException {
+		try {
+			long keyNumber = Long.parseLong( key.replaceAll( "[^0-9]", "" ) );
+			long keySpace = key.split( "\u0020" ).length - 1;
+			if( keySpace == 0 ) {
+				throw new InvalidHandshakeException( "invalid Sec-WebSocket-Key (/key2/)" );
+			}
+			long part = new Long( keyNumber / keySpace );
+			return new byte[]{ (byte) ( part >> 24 ), (byte) ( ( part << 8 ) >> 24 ), (byte) ( ( part << 16 ) >> 24 ), (byte) ( ( part << 24 ) >> 24 ) };
+		} catch ( NumberFormatException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)